src/api/fuel.ts

```diff
@@ -0,0 +1,535 @@
+import {
+  SignTransactionConfig,
+  SignTransactionResponse,
+  User
+} from 'universal-authenticator-library';
+import {
+  AnyTransaction,
+  APIClient,
+  Name,
+  NameType,
+  PackedTransaction,
+  PermissionLevel,
+  Serializer,
+  Signature,
+  SignedTransaction,
+  Transaction
+} from '@greymass/eosio';
+import { AnchorUser } from 'ual-anchor';
+import { getChain } from 'src/config/ConfigManager';
+import { Dialog } from 'quasar';
+
+// The maximum fee per transaction this script is willing to accept
+const maxFee = 0.005;
+
+// expire time in millisec
+const expireSeconds = 3600;
+
+const chain = getChain();
+const client = new APIClient({
+  url: chain.getHyperionEndpoint()
+});
+
+const fuelrpc = chain.getFuelRPCEndpoint();
+const resourceProviderEndpoint = `${fuelrpc.protocol}://${fuelrpc.host}:${fuelrpc.port}/v1/resource_provider/request_transaction`;
+
+// Auxiliar interfaces
+interface ResourceProviderResponse {
+  code: number;
+  data: ResponseData;
+}
+
+interface ResponseData {
+  request: [string, SignedTransaction];
+  signatures: Signature[];
+  version: number;
+  costs?: { ram: string };
+}
+
+interface SignedTransactionResponse extends SignTransactionResponse {
+  transaction: { signatures: Signature[] };
+}
+
+// Wrapper for the user to intersect the signTransaction call
+export class FuelUserWrapper extends User {
+  constructor(private user: User) {
+    super();
+  }
+
+  async signTransaction(
+    originalTransaction: AnyTransaction,
+    originalconfig?: SignTransactionConfig
+  ): Promise<SignTransactionResponse> {
+    try {
+      // Retrieve transaction headers
+      const info = await client.v1.chain.get_info();
+      const header = info.getTransactionHeader(expireSeconds);
+
+      // collect all contract abis
+      const abi_promises = originalTransaction.actions.map((a) =>
+        client.v1.chain.get_abi(a.account)
+      );
+      const responses = await Promise.all(abi_promises);
+      const abis = responses.map((x) => x.abi);
+      const abis_and_names = originalTransaction.actions.map((x, i) => ({
+        contract: x.account,
+        abi: abis[i]
+      }));
+
+      // create complete well formed transaction
+      const transaction = Transaction.from(
+        {
+          ...header,
+          actions: originalTransaction.actions
+        },
+        abis_and_names
+      );
+
+      // Pack the transaction for transport
+      const packedTransaction = PackedTransaction.from({
+        signatures: [],
+        packed_context_free_data: '',
+        packed_trx: Serializer.encode({ object: transaction })
+      });
+
+      const signer = PermissionLevel.from({
+        actor: (await this.user.getAccountName()) as NameType,
+        permission: this.requestPermission
+      });
+
+      // Submit the transaction to the resource provider endpoint
+      const cosigned = await fetch(resourceProviderEndpoint, {
```

Comment:
I'm not sure where it belongs, but before attempting to use a resource provider for cosigning here, it might be worth checking what chain the user is on and ensuring that Fuel is available on that chain. 

For example, we offer Fuel on Telos - but not on Telos Testnet. So this request is only ever going to succeed if the appropriate chain is in use.

Reply:
Yes, indeed. This `resourceProviderEndpoint` value is taken from a fuel-specific pre-checked field with the endpoint for every supported blockchain (as shown in the following code):

https://github.com/telosnetwork/open-block-explorer/blob/889c4ab1158441fa69bb965c086ed4d7ba13deec/src/api/fuel.ts#L35-L36

Currently, we only have the endpoints for telos and eos:

Telos:
https://github.com/telosnetwork/open-block-explorer/blob/889c4ab1158441fa69bb965c086ed4d7ba13deec/src/config/chains/telos/index.ts#L36-L40

EOS:
https://github.com/telosnetwork/open-block-explorer/blob/889c4ab1158441fa69bb965c086ed4d7ba13deec/src/config/chains/eos/index.ts#L21-L25

@aaroncox, if you do support any other antelope.io blockchain feel free to contact me to provide the missing endpoints, or if you prefer, you can create a very small PR adding it to the corresponding blockchain config file:

`/src/config/chains/<network-name>/index.ts`

Reply:
Currently Fuel supports:

- Telos
- EOS
- WAX
- Jungle 3
- Jungle 4